presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/EquivalenceClassProperty.java

```diff
@@ -0,0 +1,351 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.expressions.LogicalRowExpressions;
+import com.facebook.presto.metadata.BuiltInFunctionHandle;
+import com.facebook.presto.spi.function.FunctionHandle;
+import com.facebook.presto.spi.function.Signature;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.common.function.OperatorType.EQUAL;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Represents classes of equivalent variable and constants references that
+ * hold for a final or intermediate result set produced by a PlanNode.
+ * Variable and constant references are made equivalent via predicate application.
+ * <p>
+ * Each equivalence class is represented by a head. The head is carefully chosen
+ * to be a the member with the strongest binding (i.e. a constant).
+ * <p>
+ * Note that the equivalence class property does not store trivial equivalence
+ * classes with only one member. All variable or constant references are
+ * considered to be in their own virtual equivalence class until combined
+ * into stored larger classes by predicate application.
+ */
+public class EquivalenceClassProperty
+{
+    private final Map<RowExpression, RowExpression> equivalenceClassHeads;
+    private final Map<RowExpression, List<RowExpression>> equivalenceClasses;
+    private boolean updated;
```

Comment:
What is this variable used for? I don't see it consumed anywhere.

Reply:
It is consumed by LogicalPropertiesImpl in FilterBuilder and JoinBuilder classes.